xapian-letor/tests/api_letor.cc

```diff
@@ -180,11 +440,147 @@ DEFINE_TESTCASE(listnet_ranker, generated)
 		   ranker.score(qrel, "", "ListNet_Ranker",
 				"scorer_output.txt", 10));
     ranker.score(query, qrel, "ListNet_Ranker", "ndcg_output.txt", 10);
-    ranker.score(query, qrel, "ListNet_Ranker", "err_output.txt", 10, "ERRScore");
+    ranker.score(query, qrel, "ListNet_Ranker", "err_output.txt", 10,
+								"ERRScore");
 
     return true;
 }
 
+
+DEFINE_TESTCASE(listnet_ranker_one_file, generated)
+{
+    Xapian::ListNETRanker ranker;
+    TEST_EXCEPTION(Xapian::FileNotFoundError, ranker.train_model(""));
+    string db_path = get_database_path("apitest_listnet_ranker5",
+				       db_index_one_document);
+    Xapian::Enquire enquire((Xapian::Database(db_path)));
+    enquire.set_query(Xapian::Query("tigers"));
+    Xapian::MSet mymset = enquire.get_mset(0, 10);
+    string data_directory = test_driver::get_srcdir() + "/testdata/";
+    string query = data_directory + "queryone.txt";
+    string qrel = data_directory + "qrelone.txt";
+    string training_data = data_directory + "training_data_one_document.txt";
+    ranker.set_database_path(db_path);
+    TEST_EQUAL(ranker.get_database_path(), db_path);
+    ranker.set_query(Xapian::Query("tigers"));
+    ranker.train_model(training_data);
+    Xapian::docid doc1 = *mymset[0];
+    ranker.rank(mymset);
+    TEST_EQUAL(doc1, *mymset[0]);
+    mymset = enquire.get_mset(0, 10);
+    ranker.train_model(training_data, "ListNet_Ranker");
+    ranker.rank(mymset, "ListNet_Ranker");
+    TEST_EQUAL(doc1, *mymset[0]);
+    TEST_EXCEPTION(Xapian::LetorInternalError,
+		   ranker.score(query, qrel, "ListNet_Ranker",
+				"scorer_output1.txt", 10, ""));
+    TEST_EXCEPTION(Xapian::FileNotFoundError,
+		   ranker.score("", qrel, "ListNet_Ranker",
+				"scorer_output1.txt", 10));
+    TEST_EXCEPTION(Xapian::FileNotFoundError,
+		   ranker.score(qrel, "", "ListNet_Ranker",
+				"scorer_output1.txt", 10));
+    ranker.score(query, qrel, "ListNet_Ranker", "ndcg_output1.txt", 10);
+    ranker.score(query, qrel, "ListNet_Ranker", "err_output1.txt", 10,
+								"ERRScore");
+
+    return true;
+}
+
```

Comment:
> One way to go about it would be to add a boolean value "set_label" in FeatureVector which would be set by default as False and will only be set True when the label has been assigned. This bool can be accessed by the Scorers which can then extract the labels which have been set and give the corresponding score.

Rather than have a flag member variable to say if the `label` member variable is valid or not, I'd suggest setting `label` to some invalid value (if `0.0` isn't a valid label, that makes a good default for a double as it's typically cheap to initialise to - all zero bits in the commonly used IEEE floating point representation - and cheap to check against; a negative value is another option; or `HUGE_VAL` if both there are both positive and negative valid values).

If you have a thing and a flag to say if the thing is valid, then you have two variables whose states are entwined but for which some combinations of states are "impossible", and you then need to worry about what happens if there's a bug and we end up in one of those "impossible" states.  With C++17 there's [`std::optional`](http://en.cppreference.com/w/cpp/utility/optional) which provides a class to encapsulate that nicely, but we can't assume C++17 yet.  We could potentially provide an equivalent class ourselves but we don't have one currently.

Adding a flag (or using `std::optional`) also increases the size of the class a little.

When all values of the thing are valid, we will need some sort of flag to add an invalid state, but it's generally nicer to avoid a flag if we can.